Prune premature case diffs

millerdev opened the pull request:
🐡 Review by commit.

millerdev committed:
Limit number of forms per case diff batch

The case diff process was consuming huge amounts (16GB) of memory. This is an attempt to resolve that situation.

millerdev committed:
Log warning on begin processing diff batch

millerdev committed:
Tweak constants

Reduce max diff workers to try to limit memory by limiting the number of loaded forms and cases. Increase forms per diff to reduce chattiness of warnings about large case diffs. Theoretically the number of total loaded forms is the same, but the theoretical limit can be exceeded since a single case may have more forms than the per-diff limit.

millerdev committed:
Improve error handling on case diff process crash

millerdev committed:
Fix exit_on_error in case of return value

millerdev committed:
Log case id on diff case with many forms

millerdev committed:
Move CaseRecord and helpers to bottom of module

millerdev committed:
Refactor prep for prune premature diffs

millerdev committed:
Defer premature diffs

The total_forms number stored in the statedb can get outdated if more forms are submitted against the case between the time when the case is first seen and when it finally becomes due to diff, resulting in a premature diff. One possible disadvantage is that a case could get

dannyroberts commented on corehq/apps/couch_sql_migration/casediff.py:
```diff
@@ -611,6 +617,26 @@ def clean_break(self, value):
+    stock_forms = get_stock_forms_by_case_id(list(cases_by_id))
+    recs = [CaseRecord(c, stock_forms.get(c.case_id, [])) for c in couch_cases]
+    for case_id, total_forms, processed_forms in statedb.update_cases(recs):
+        if total_forms <= processed_forms:
```
Can you describe this to me in terms of what is the "originally calculated" number of forms (possibly calculated weeks ago) and what is the current number of forms? Or maybe this isn't where it's happening; I'm a big confused what's getting filtered out here, and whether that originally calculated number of forms is getting updated

millerdev replied:
A new value of `total_forms` is calculated in the constructor of `CaseRecord` (line 625). The case is _usually freshly loaded from couch_*. `processed_forms` is the number of forms touching this case that have been migrated so far. This is the same condition used enqueue a case to be diffed:

https://github.com/dimagi/commcare-hq/blob/f3a9efe00c916e2e25ccca9d2fc03c7900641d39/corehq/apps/couch_sql_migration/casediff.py#L150-L151

*The exception to _usually freshly loaded from couch_ is when a stale version of the case was present in the LRU cache (not a likely scenario), in which case there is a very slight chance of a false negative (no diff when there should be a diff) if the stale version of the case compares equal to the newer version in SQL, and the newer version in Couch differs. This edge case seems sufficiently unlikely to ignore.

millerdev committed:
De-six teeout.py

millerdev committed:
Disable teeout() tests on travis

Example failures:

https://api.travis-ci.org/v3/job/606077793/log.txt
======================================================================
FAIL: corehq.util.tests.test_teeout:TestTee.test_tee_output
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/util/tests/test_teeout.py", line 22, in test_tee_output
    eq(fake.stderr.getvalue(), "fail.\n")
  File "/vendor/lib/python3.6/site-packages/testil.py", line 54, in eq
    raise AssertionError(err)
AssertionError: not equal
"fail.\n/usr/local/lib/python3.6/traceback.py:310: ResourceWarning: unclosed <socket.socket fd=29, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.18.0.12', 52426), raddr=('172.18.0.2', 5984)>\n  yield tb.tb_frame, tb.tb_lineno\n/usr/local/lib/python3.6/traceback.py:310: ResourceWarning: unclosed <socket.socket fd=30, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.18.0.12', 52460), raddr=('172.18.0.2', 5984)>\n  yield tb.tb_frame, tb.tb_lineno\n/usr/local/lib/python3.6/traceback.py:310: ResourceWarning: unclosed file <_io.BufferedReader name='/mnt/commcare-hq-ro/corehq/util/tests/data/duplicate_columns.xlsx'>\n  yield tb.tb_frame, tb.tb_lineno\n"
'fail.\n'
----------------------------------------------------------------------

https://api.travis-ci.org/v3/job/545327504/log.txt
======================================================================
ERROR: corehq.util.tests.test_teeout:test_tee_output_with_KeyboardInterrupt ()
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/mnt/commcare-hq-ro/corehq/util/tests/test_teeout.py", line 42, in test_tee_output_with_KeyboardInterrupt
    eq(fake.stdout.getvalue(), "")
ValueError: I/O operation on closed file
----------------------------------------------------------------------

kaapstorm reviewed (commented):
:black_heart:
